Refactored beets/random.py  and moved into beetsplug/random.py

semohr opened the pull request:
The `beets/random.py` module was only used by the random plugin, so I moved its functions into `beetsplug/random.py` to keep core modules cleaner.


Changes:
- Moved beets/random.py functions into beetsplug/random.py
- Added typehints for better readability and tooling support
- Added additional tests for improved coverage
- General tidy up and refactor, keeping the core functionality unchanged

semohr committed:
Moved functions from random.py into random plugin. Removed random.py

semohr committed:
Added typehints and some more tests.

snejus commented on beetsplug/random.py:
```diff
@@ -64,3 +71,107 @@ def random_func(lib, opts, args):
+    return out
+
+
+def random_objs(
+    objs: Sequence[Item | Album],
+    number=1,
+    time: float | None = None,
+    equal_chance: bool = False,
+    random_gen: random.Random | None = None,
+):
```
This function definition doesn't seem to be correct given the way it's called:

```py
    objs = random_objs(
        objs, opts.album, opts.number, opts.time, opts.equal_chance
    )
```

I think `opts.album` is missing? `random_gen` argument is never used, so you can use `random.shuffle` directly.

semohr commented on beetsplug/random.py:
```diff
@@ -64,3 +71,107 @@ def random_func(lib, opts, args):
+T = TypeVar("T")
+
+
+def _take(
+    iter: Iterable[T],
+    num: int,
+) -> list[T]:
+    """Return a list containing the first `num` values in `iter` (or
+    fewer, if the iterable ends early).
+    """
+    return list(islice(iter, num))
```
I think this `_take` function is the only one I modified, sine it was obviously highly inefficient and I was wondering what the initial code was even doing.

snejus replied:
I'd be happy to have a small refactor as part of this PR if you're up for it! 😅

semohr committed:
Overall refactor of random plugin. Added length property to albums.

snejus reviewed (changes requested):
A couple of more comments. I found that unused default arguments are still present, so I unresolved my previous two comments where I commented about them.

semohr commented on beetsplug/random.py:
```diff
@@ -64,3 +62,95 @@ def random_func(lib, opts, args):
+
+def random_objs(
+    objs: Sequence[T],
+    number: int = 1,
+    time_minutes: float | None = None,
+    equal_chance: bool = False,
+    random_gen: random.Random | None = None,
```
Let's do parameter parsing as early as possible. Am not a big fan passing unknown/untyped values.

We use the default parameters in the tests all over the place. I would keep them just for usability even tho they are not used in the main function call. Doesn't really hurt imo.

semohr commented on beetsplug/random.py:
```diff
@@ -64,3 +62,95 @@ def random_func(lib, opts, args):
+    def get_attr(obj: T) -> Any:
+        try:
+            return key(obj)
+        except AttributeError:
+            return NOT_FOUND_SENTINEL
```
This can occur if we reference a field that might not be defined on an object. In practice, this is very unlikely and currently not possible given how the function is invoked. However, we do have a test for it, as it is technically within the realm of possibility.

I’d prefer to keep the function as generic as possible, since this flexibility could be useful if we decide to extend the random plugin with additional features in the future.

semohr commented on beetsplug/random.py:
```diff
@@ -1,37 +1,35 @@
-# Permission is hereby granted, free of charge, to any person obtaining
-# a copy of this software and associated documentation files (the
-# "Software"), to deal in the Software without restriction, including
-# without limitation the rights to use, copy, modify, merge, publish,
```
You have seen that this was not the standard copyright notice?

snejus replied:
How come? 
<img width="747" height="201" alt="image" src="https://github.com/user-attachments/assets/470c94b7-ab03-4395-8954-38d51b8643a2" />

semohr commented on beetsplug/random.py:
```diff
@@ -64,3 +62,95 @@ def random_func(lib, opts, args):
+        except AttributeError:
+            return NOT_FOUND_SENTINEL
+
+    groups: dict[Any, list[T]] = {
```
At the moment that is true but as I said in another comment. I’d prefer to keep the function as generic as possible, since this flexibility could be useful if we decide to extend the random plugin with additional features in the future.

Beeing able to define any field is a nice feature even if unused atm.

semohr commented on beetsplug/random.py:
```diff
@@ -64,3 +71,107 @@ def random_func(lib, opts, args):
+
+def _equal_chance_permutation(
+    objs: Sequence[Item | Album],
+    field: str = "albumartist",
+    random_gen: random.Random | None = None,
```
See other comments. Also we need to keep the random gen variable to ensure reproducible tests here.

semohr committed:
Cached property for length & forgot sorted.

semohr commented on beetsplug/random.py:
```diff
@@ -64,3 +62,95 @@ def random_func(lib, opts, args):
+    groups: dict[Any, list[T]] = {
+        NOT_FOUND_SENTINEL: [],
+    }
+    for k, values in groupby(objs, key=get_attr):
```
See other comment on fields. Also seems like a nit not really worth our time tbh ;)

semohr commented on beetsplug/random.py:
```diff
@@ -1,37 +1,35 @@
+
+    from beets.library import LibModel, Library
+
+    T = TypeVar("T", bound=LibModel)
```
> For clarity and to help type-checking it would be ideal to define their types under LibModel

Yeah would be nice to identify more shared functionalities. I think this is out of scope here tho.

snejus replied:
Apologies if I wasn't clear - I don't see why are we using `TypeVar` here instead of the simpler `LibModel` alternative.

As I said above, this relates to the `length` property, which is currently not resolved properly (unknown for `Item`) and thus isn't type-checked:

```py
def _take_time(
    iter: Iterable[T],
    secs: float,
) -> Iterable[T]:
    """Return a list containing the first values in `iter`, which should
    be Item or Album objects, that add up to the given amount of time in
    seconds.
    """
    total_time = 0.0
    for obj in iter:
        length = obj.length
        reveal_type(obj.length)  # mypy: Revealed type is "Union[Any, builtins.float]"
        if total_time + length <= secs:
            yield obj
            total_time += length
```

Instead, defining `LibModel.length` type

```py
class LibModel(dbcore.Model["Library"]):
    """Shared concrete functionality for Items and Albums."""

    # Config key that specifies how an instance should be formatted.
    _format_config_key: str
    path: bytes
    length: float
```

and replacing the type var with `LibModel` fixes it:

```py
def _take_time(
    iter: Iterable[LibModel],
    secs: float,
) -> Iterable[LibModel]:
    """Return a list containing the first values in `iter`, which should
    be Item or Album objects, that add up to the given amount of time in
    seconds.
    """
    total_time = 0.0
    for obj in iter:
        length = obj.length
        reveal_type(obj.length)  # mypy: Revealed type is "builtins.float"
        if total_time + length <= secs:
            yield obj
            total_time += length
```

Additionally, this prevents `Item` or `Album` from returning any other type for this property.

snejus replied:
Shall I commit the replacement of `T` by `LibModel` or are you happy doing it yourself?

snejus commented on beetsplug/random.py:
```diff
@@ -64,3 +62,95 @@ def random_func(lib, opts, args):
+      selected, regardless of how many tracks they have.
+    - random_gen: An optional random generator to use for shuffling.
+    """
+    rand: random.Random = random_gen or random.Random()
```
The issue with type hints disappears once it's removed from function parameters and we use `random` module directly in those three places.

Regarding

> It is mostly the same thing. The random module is a hidden random.Random instance.

> Also we need to keep the random gen variable to ensure reproducible tests here.

Not sure if I'm missing something but seeding the `random` module globally once seems to be more straightforward and reliable than depending on a seeded `random.Random` instance being passed across the functions?

snejus replied:
I did edit the source code before I commented to double-check that it was valid, so I committed and pushed it now.

snejus committed:
Replace random.Random with random module

semohr committed:
Added an option to define the field to use for equal chance sampling

snejus committed:
Add annotation for LibModel.length property

snejus reviewed (changes requested):
Since you've added `field` argument, I added a couple of related comments. This should be mentioned in the changelog as well.

semohr committed:
Readded licence. Removed last legacy occurrences of `artist` and
replaced them with `field`. Removed unnecessary default parameters where
applicable.

semohr committed:
Enhanced documentation of random plugin.

snejus commented on beetsplug/random.py:
```diff
@@ -57,10 +54,105 @@ def random_func(lib, opts, args):
+def _take_time(
+    iter: Iterable[LibModel],
+    secs: float,
+) -> Iterable[LibModel]:
```
Now I see... I'm officially a muppet

snejus reviewed (changes requested):
I gave this a test and found two bugs:

1. List fields are not handled
```sh
$ beet random --field artists -e
Traceback (most recent call last):
  File "/home/sarunas/.local/bin/beet", line 8, in <module>
    main()
  File "/home/sarunas/repo/beets/beets/ui/__init__.py", line 1629, in main
    _raw_main(args)
  File "/home/sarunas/repo/beets/beets/ui/__init__.py", line 1608, in _raw_main
    subcommand.func(lib, suboptions, subargs)
  File "/home/sarunas/repo/beets/beetsplug/random.py", line 39, in random_func
    for obj in random_objs(
  File "/home/sarunas/repo/beets/beetsplug/random.py", line 112, in _equal_chance_permutation
    groups[k] = list(values)
TypeError: unhashable type: 'list'```
```

2. items aren't sorted before grouping in equal chance permutation. See `beet list` output:

```sh
$ beet list path::aaa title- -f '$album: $artist - $title'
BLAH!!!! malakaa!! todotiernos de mierdaaa: DER SANDMANN - Triple Masala
BLAH!!!! malakaa!! todotiernos de mierdaaa: DER SANDMANN - Machete
BLAH!!!! malakaa!! todotiernos de mierdaaa: DER SANDMANN - Lost in Time
BLAH!!!! malakaa!! todotiernos de mierdaaa: DER SANDMANN - Freaks at Work
BLAH!!!! malakaa!! todotiernos de mierdaaa: DER SANDMANN - der killer wahl 185bpm
RAVE SLUTZ: Fallen Shrine & dj Christian NXC - DEEEJAAAY
BLAH!!!! malakaa!! todotiernos de mierdaaa: DER SANDMANN - Blah
Netikras albumas: Karpiz - AAA
```

vs

```sh
$ beet random path::aaa title- -f '$album: $artist - $title' --field album -e -n10
RAVE SLUTZ: Fallen Shrine & dj Christian NXC - DEEEJAAAY
Netikras albumas: Karpiz - AAA
BLAH!!!! malakaa!! todotiernos de mierdaaa: DER SANDMANN - Blah
```

I expected to see the same items but in random order.

I'm keen to finally wrap this up, so feel free to just paste this fix:

```diff
diff --git a/beetsplug/random.py b/beetsplug/random.py
index 9714c8e53..aa65dd5d5 100644
--- a/beetsplug/random.py
+++ b/beetsplug/random.py
@@ -19,4 +19,4 @@
 from itertools import groupby, islice
-from operator import attrgetter
-from typing import TYPE_CHECKING, Any
+from operator import methodcaller
+from typing import TYPE_CHECKING
 
@@ -86,5 +86,2 @@ def commands
 
-NOT_FOUND_SENTINEL = object()
-
-
 def _equal_chance_permutation(
@@ -97,22 +94,12 @@ def _equal_chance_permutation
     # Group the objects by field so we can sample from them.
-    key = attrgetter(field)
-
-    def get_attr(obj: LibModel) -> Any:
-        try:
-            return key(obj)
-        except AttributeError:
-            return NOT_FOUND_SENTINEL
-
-    sorted(objs, key=get_attr)
-
-    groups: dict[str | object, list[LibModel]] = {
-        NOT_FOUND_SENTINEL: [],
-    }
-    for k, values in groupby(objs, key=get_attr):
-        groups[k] = list(values)
-        # shuffle in category
-        random.shuffle(groups[k])
-
-    # Remove items without the field value.
-    del groups[NOT_FOUND_SENTINEL]
+    get_attr = methodcaller("get", field)
+
+    groups = {}
+    for k, values in groupby(sorted(objs, key=get_attr), key=get_attr):
+        if k is not None:
+            vals = list(values)
+            # shuffle in category
+            random.shuffle(vals)
+            groups[str(k)] = vals
+
     while groups:
```

snejus committed:
Fix grouping for list fields and stabilize equal-chance order

- Handle list-valued fields when grouping for --field/--equal-chance to avoid
  "TypeError: unhashable type: 'list'" (e.g., artists).
- Sort items by the grouping key before building groups so equal-chance
  permutation preserves the same item set as `beet list`, only randomized.
